File: software/mosek/7/tools/examples/python/mioinitsol.py
##
#    Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#    File:    mioinitsol.py
#
#    Purpose:  Demonstrates how to solve a small mixed
#              integer linear optimization problem using the MOSEK Python API.
##

import sys

import mosek

# If numpy is installed, use that, otherwise use the 
# Mosek's array module.
try:
    from numpy import array,zeros,ones
except ImportError:
    from mosek.array import array, zeros, ones

# Since the actual value of Infinity is ignores, we define it solely
# for symbolic purposes:
inf = 0.0


# Define a stream printer to grab output from MOSEK
def streamprinter(text):
    sys.stdout.write(text)
    sys.stdout.flush()


# We might write everything directly as a script, but it looks nicer
# to create a function.
def main ():
    # Make a MOSEK environment
    env = mosek.Env ()
    # Attach a printer to the environment
    env.set_Stream (mosek.streamtype.log, streamprinter)

    # Create a task
    task = env.Task(0,0)
    # Attach a printer to the task
    task.set_Stream (mosek.streamtype.log, streamprinter)

       
    bkc  = [ mosek.boundkey.up ]
    blc  = [ -inf,             ]
    buc  = [ 2.5               ] 

    bkx  = [ mosek.boundkey.lo,
             mosek.boundkey.lo,
             mosek.boundkey.lo,
             mosek.boundkey.lo ]
    
    blx  = [0.0, 0.0, 0.0, 0.0 ]
    bux  = [ inf, inf,  inf,  inf ]

    c    = [ 7.0, 10.0, 1.0, 5.0 ]

    asub = [  0,   0,   0,   0   ]
    acof = [  1.0, 1.0, 1.0, 1.0]

    ptrb = [ 0, 1, 2, 3 ]
    ptre = [ 1, 2, 3, 4 ]

    numvar = len(bkx)
    numcon = len(bkc)

    # Input linear data
    task.inputdata(numcon,numvar,
                   c,0.0,
                   ptrb, ptre, asub, acof,
                   bkc,  blc,  buc,
                   bkx,  blx,  bux)

    # Input objective sense
    task.putobjsense(mosek.objsense.maximize)

    # Define variables to be integers
    task.putvartypelist([ 0, 1, 2 ],
                        [ mosek.variabletype.type_int,
                          mosek.variabletype.type_int,
                          mosek.variabletype.type_int])
    
    # Construct an initial feasible solution from the
    #     values of the integer valuse specified 
    task.putintparam(mosek.iparam.mio_construct_sol,
                     mosek.onoffkey.on);

    # Assign values 0,2,0 to integer variables. Important to
    # assign a value to all integer constrained variables. 
    task.putxxslice(mosek.soltype.itg,0,3,[0.0, 2.0, 0.0])

    # Optimize
    task.optimize()

    # Did mosek construct a feasible initial solution ?
    if task.getintinf(mosek.iinfitem.mio_construct_solution) > 0:
        print("Objective value of constructed integer solution: %-24.12e" % task.getdouinf(mosek.dinfitem.mio_construct_solution_obj))
    else:
        print("Intial integer solution construction failed.");    

    if task.solutiondef(mosek.soltype.itg):

        # Output a solution
        xx = zeros(numvar, float)
        task.getxx(mosek.soltype.itg, xx)
        print("Integer optimal solution")
        for j in range(0,numvar) :
            print("\tx[%d] = %e" % (j,xx[j]))
    else:
        print("No integer solution is available.")

# call the main function
try:
    main ()
except mosek.Exception as e:
    print ("ERROR: %s" % str(e.errno))
    if e.msg is not None:
        print ("\t%s" % e.msg)
    sys.exit(1)
except:
    import traceback
    traceback.print_exc()
    sys.exit(1)

File: software/mosek/7/tools/examples/java/mioinitsol.java
package com.mosek.example;

/*
   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.

   File:      mioinitsol.java

   Purpose:   Demonstrates how to solve a MIP with a start guess.

   Syntax:    mioinitsol
 */

 
import mosek.*;

class msgclass extends mosek.Stream {
    public msgclass ()
    {
        super ();
    }

    public void stream (String msg)
    {
        System.out.print (msg);
    }
}
 
public class mioinitsol
{
    public static void main (String[] args)
    {
        mosek.Env
            env = null;
        mosek.Task
            task = null;
        // Since the value infinity is never used, we define
        // 'infinity' symbolic purposes only
        double
            infinity = 0;

        int numvar = 4;
        int numcon = 1;

        double[] c = { 7.0, 10.0, 1.0, 5.0 };
        
        mosek.Env.boundkey[] bkc = {mosek.Env.boundkey.up};
        double[] blc = {-infinity};
        double[] buc = {2.5};
        mosek.Env.boundkey[] bkx
                  = {mosek.Env.boundkey.lo,
                     mosek.Env.boundkey.lo,
                     mosek.Env.boundkey.lo,
                     mosek.Env.boundkey.lo};
        double[] blx = {0.0,
                        0.0,
                        0.0,
                        0.0};
        double[] bux = {infinity,
                        infinity,
                        infinity,
                        infinity};
        
        int[]    ptrb   = {0, 1, 2, 3};
        int[]    ptre   = {1, 2, 3, 4};
        double[] aval   = {1.0, 1.0, 1.0, 1.0};
        int[]    asub   = {0,   0,   0,   0  };
        int[]    intsub = {0, 1, 2};
        mosek.Env.variabletype[] inttype={mosek.Env.variabletype.type_int,
                                          mosek.Env.variabletype.type_int,
                                          mosek.Env.variabletype.type_int};
        double[]    intxx={0.0,2,0,0,0};

        try{
            // Make mosek environment. 
            env  = new mosek.Env ();
            // Create a task object linked with the environment env.
            task = new mosek.Task (env, 0, 0);
            // Directs the log task stream to the user specified
            // method task_msg_obj.print
            msgclass task_msg_obj = new msgclass ();
            task.set_Stream (mosek.Env.streamtype.log,task_msg_obj);

            task.inputdata(numcon,numvar,
                           c, 0.0,
                           ptrb, ptre,
                           asub, aval,
                           bkc, blc, buc,
                           bkx, blx, bux);
                            
            task.putvartypelist(intsub,inttype);

            /* A maximization problem */ 
            task.putobjsense(mosek.Env.objsense.maximize);
                
            // Construct an initial feasible solution from the
            //     values of the integer valuse specified 
            task.putintparam(mosek.Env.iparam.mio_construct_sol,
                             mosek.Env.onoffkey.on.value);
        
            // Set status of all variables to unknown 
            //task.makesolutionstatusunknown(mosek.Env.soltype.itg);

            // Assign values 1,1,0 to integer variables 
            task.putxxslice(mosek.Env.soltype.itg, 0, 3, intxx);

            // solve   
            task.optimize(); 
        }
        catch (mosek.Exception e)
        /* Catch both Error and Warning */
        {
            System.out.println ("An error was encountered");
            System.out.println (e.getMessage ());
            throw e;
        }
        finally
        {
            if (task != null) task.dispose ();
            if (env  != null)  env.dispose ();
        }        
    }
}
